### PR TITLE
Transforms end to end proof of concept

### PR DESCRIPTION
**Pull request by BrennaCourtney**
### Summary of changes 🔍 
Author and view transforms in 3DScenes
image
image

Design document (with progress, notes, etc): https://microsoft-my.sharepoint.com/:w:/r/personal/t-brcourtney_microsoft_com/_layouts/15/doc2.aspx?sourcedoc=%7BF3A652E0-3049-42CE-9FFC-4ACD38FA9564%7D&file=Mesh%20Transformations%20Design%20Doc.docx&nav=eyJjIjoyMDY1NDkyMDc4fQ&action=default&mobileredirect=true&share=IQHgUqbzSTDOQp_8Ss04-pVkAaEgTWAsSWVRN7tQw0x2NGM

Known challenges/areas for investigation:
- Assigning transforms to multiple elements with a behavior
- Assigning different transforms to different value ranges
- Merging of Status and Alerts tab -> Visuals tab (with transforms inside?)
- Modifying schema so that transforms can be strongly typed and do not exist on "extensionProperties"
- "Enacting" transforms rather than just setting an element to an XYZ position/rotation -- that is, "adding" XYZ rather than "setting" (will make transforms more transferable to other elements)
- Reading existing transforms assigned to current element+behavior in Build Mode (right now, just resets, so you need to set the transform again -- would require reading config file)
- Colored/highlighted meshes sometimes use clones, and these clones (especially if the element is already animated) do not always move with the transformed element
- Radians -> degrees preferred in text fields
- "Revert to original (position/rotation)" button in Build mode

### Testing 🧪
 Main flow is the following:
- Enter the Alerts tab on a behavior (preferably with only one non-animated element)
- Manipulate the element using the gizmo or the text fields
- "Update behavior" to save to the config file
- Trigger transforms behavior in viewer side. May take a few seconds to see the effect (that is, until the data is "pulled")

Demo here: https://msit.microsoftstream.com/video/1a9d0840-98dc-ae81-ce7e-f1ed13678ba8?list=studio

### Checklist ✔️
- [ ] Linked associated issue (if present)
- [ ] Added relevant labels
- [ ] Requested developer reviews
- [ ] Request UI review from design / PM
- [ ] Verify all code checks are passing

### CODE REVIEW

File: src/Components/3DV/SceneView.tsx [hunk 1]
```diff
@@ -79,7 +79,7 @@ import { useTranslation } from 'react-i18next';
 
 export const showFpsCounter = false;
 const debugBabylon = false;
-const debugLogging = false;
+const debugLogging = true;
```
Comment (iremgokce) [inline src/Components/3DV/SceneView.tsx:82]: Lets set logging back to false if not intended

File: src/Components/ADT3DSceneBuilder/Internal/Behaviors/Internal/AlertsTab.tsx [hunk 2]
```diff
@@ -69,11 +73,61 @@ const AlertsTab: React.FC = () => {
...
     const alertVisualStateRef = useRef<IExpressionRangeVisual>(
         getAlertFromBehavior(behaviorFormState.behaviorToEdit) ||
             getDefaultAlertVisualWithId()
     );
 
+    const gizmoElementRef = useRef<TransformedElementItem>(null);
+
+    useEffect(() => {
+        if (selectedElements.length > 0) {
+            console.log('selectedElements: ', selectedElements);
```
Comment (iremgokce) [inline src/Components/ADT3DSceneBuilder/Internal/Behaviors/Internal/AlertsTab.tsx:87]: lets remove console.logs here and below

File: src/Components/ADT3DSceneBuilder/Internal/Behaviors/Internal/AlertsTab.tsx [hunk 3]
```diff
@@ -69,11 +73,61 @@ const AlertsTab: React.FC = () => {
...
+                    meshIds.push(objectID);
+                });
+                setGizmoElementItem(gizmoElementRef.current); // just adding one gizmoElement so far ...
+                console.log('gizmo element ref: ', gizmoElementRef.current);
+            }
+        }
+        return () => {
+            // clean up gizmo after exiting tab
+            setGizmoElementItem({ parentMeshId: null, meshIds: null });
+            // setGizmoElementItem(null);
```
Comment (iremgokce) [inline src/Components/ADT3DSceneBuilder/Internal/Behaviors/Internal/AlertsTab.tsx:110]: remove comment please

File: src/Components/BabylonSandbox/BabylonSandbox.tsx [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+>();
+
+/**
+ * Sets up the camera for scene.
+ */
+function setupCamera(
+    scene: BABYLON.Scene,
+    arcRotate: boolean,
+    center?: BABYLON.Vector3,
+    radius?: number
```
Comment (iremgokce) [inline src/Components/BabylonSandbox/BabylonSandbox.tsx:27]: lets fix linting warnings like this one by running `>npm run lint:check` and `>npm run lint` to fix those
```suggestion
    _radius?: number
```

File: src/Components/BabylonSandbox/BabylonSandbox.tsx [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+): BABYLON.ArcRotateCamera | BABYLON.FreeCamera {
+    // Creates, angles, distances and targets the camera
+    const camera = arcRotate
+        ? new BABYLON.ArcRotateCamera(
+              'Camera',
+              0,
+              Math.PI / 2.5,
+              //   radius ? radius : 33,
```
Comment (iremgokce) [inline src/Components/BabylonSandbox/BabylonSandbox.tsx:35]: remove comment please

File: src/Components/BabylonSandbox/BabylonSandbox.tsx [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    const [canvasId] = useState(createGUID());
+    const [scene, setScene] = useState<BABYLON.Scene>();
+    const engineRef = useRef<BABYLON.Engine>(null);
+    const pickedMeshRef = useRef<BABYLON.AbstractMesh>(null);
+    const gizmoManagerRef = useRef<BABYLON.GizmoManager>(undefined);
+
+    // setting up engine based on the canvas & scene based on engine
+    useEffect(() => {
+        const canvas = document.getElementById(canvasId) as HTMLCanvasElement;
+        console.log(canvas);
```
Comment (iremgokce) [inline src/Components/BabylonSandbox/BabylonSandbox.tsx:95]: lets remove this console.log too

File: src/Components/BabylonSandbox/BabylonSandbox.tsx [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+            gizmoManagerRef.current = new BABYLON.GizmoManager(scene);
+            const gizmoManager = gizmoManagerRef.current;
+            gizmoManager.usePointerToAttachGizmos = false;
+            gizmoManager.rotationGizmoEnabled = true;
+            gizmoManager.positionGizmoEnabled = true;
+            // gizmoManager.gizmos.boundingBoxGizmo.ignoreChildren = true;
+
+            BABYLON.SceneLoader.ImportMeshAsync(
+                '',
+                // 'https://assets.babylonjs.com/meshes/',
```
Comment (iremgokce) [inline src/Components/BabylonSandbox/BabylonSandbox.tsx:117]: lets remove comments

File: src/Components/BabylonSandbox/BabylonSandbox.tsx [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+                // 'https://dl.dropbox.com/s/s1p66hldqurou0w/Xwing_rig.glb',
+                'https://dl.dropbox.com/s/2cq4fnsg8nqxckg/TruckBoxesEnginesPastmachine.gltf',
+                // 'https://dl.dropbox.com/s/t1wepbusypqevn4/mercedes.glb',
+                // 'both_houses_scene.babylon'
+                // 'Xwing_rig.glb'
+                'TruckBoxesEnginesPastmachine.gltf'
+                // 'mercedes.glb'
+            ).then((result) => {
+                const meshes = result.meshes;
+                const bbox = getBoundingBox(meshes);
```
Comment (iremgokce) [inline src/Components/BabylonSandbox/BabylonSandbox.tsx:127]: fix please linting issues please
```suggestion
               
```

File: src/Components/BabylonSandbox/BabylonSandbox.tsx [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+                // const center = bbox.boundingBox.centerWorld;
```
Comment (iremgokce) [inline src/Components/BabylonSandbox/BabylonSandbox.tsx:128]: lets clean the comments please

File: src/Components/BabylonSandbox/BabylonSandbox.tsx [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+                            positionTB.text = text;
+                        }
+                    }
+                )
+            );
+
+            //////////// GUI STUFF //////////////////////////////////////////
+
+            scene.onPointerDown = function castRay() {
+                // scene.onKeyboardObservable.clear();
```
Comment (iremgokce) [inline src/Components/BabylonSandbox/BabylonSandbox.tsx:211]: remove the comments please

File: src/Components/BabylonSandbox/BabylonSandbox.tsx [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+
+                const hit = scene.pickWithRay(ray);
+
+                if (hit.pickedMesh) {
+                    pickedMeshRef.current = hit.pickedMesh;
+                    // console.log(hit.pickedMesh);
+                    // setGizmoManager(new BABYLON.GizmoManager(scene));
+                    pickedMeshRef.current.rotationQuaternion = null;
+                    gizmoManager.attachToMesh(pickedMeshRef.current);
+                    console.log('gizmo?');
```
Comment (iremgokce) [inline src/Components/BabylonSandbox/BabylonSandbox.tsx:228]: remove please

File: src/Components/BabylonSandbox/BabylonSandbox.types.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    /**
+     * SubComponent styles.
+     */
+    subComponentStyles?: IBabylonSandboxSubComponentStyles;
+}
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface IBabylonSandboxSubComponentStyles {}
+
+// export interface TransformMeshItem {
```
Comment (iremgokce) [inline src/Components/BabylonSandbox/BabylonSandbox.types.ts:28]: lets remove unused comments

File: src/Models/Hooks/useRuntimeSceneData.ts [hunk 13]
```diff
@@ -71,6 +73,91 @@ export const useRuntimeSceneData = (
                 sceneVisual: SceneVisual;
                 sceneViewBadge: SceneViewBadge;
             }> = [];
+
+            // for each scene visual retrieve the TRANSFORMED mesh ids and update it in the scene visual
+            // if they are triggered by the element's behaviors and currently active
+            sceneVisuals.forEach((sceneVisual) => {
+                sceneVisual.transformedElementItems = []; // initialize
+
+                // const transformedMeshItems: Array<CustomMeshItem> = []; --> an array of all the transformed meshes
```
Comment (iremgokce) [inline src/Models/Hooks/useRuntimeSceneData.ts:82]: lets clean the comments in this file too

File: src/Models/Hooks/useRuntimeSceneData.ts [hunk 14]
```diff
@@ -71,6 +73,91 @@ export const useRuntimeSceneData = (
...
+                                            position: {
+                                                x: extensionProperties.xPos as number,
+                                                y: extensionProperties.yPos as number,
+                                                z: extensionProperties.zPos as number
+                                            }
+                                        };
+                                        // iterate through each mesh in the element and add its meshId to a
+                                        // TransformedElement object and assign the transform (as decided by
+                                        // extensionProperties) to that TransformedElement
+                                        // problem: who is the parent? for now, we can just do the first one ...
```
Comment (iremgokce) [inline src/Models/Hooks/useRuntimeSceneData.ts:130]: we can add TODO: just to mark this for future reference 
```suggestion
                                        // TODO: problem: who is the parent? for now, we can just do the first one ...
```

File: src/Models/Hooks/useRuntimeSceneData.ts [hunk 15]
```diff
@@ -71,6 +73,91 @@ export const useRuntimeSceneData = (
...
+                                                );
+                                            }
+                                        );
+                                        // for now just set parent to be first mesh
+                                        transformedElementItem.parentMeshId =
+                                            transformedElementItem.meshIds[0];
+                                        transformedElementItem.transform = deepCopy(
+                                            transform
+                                        );
+                                        // console.log(transformedElementItem);
```
Comment (iremgokce) [inline src/Models/Hooks/useRuntimeSceneData.ts:149]: lets remove the unused comments :)

**Comment by darsney** [comment]
abandoning.  this branch is now saved at[ transformExamples/integratingIntoScenesApp ](https://github.com/microsoft/iot-cardboard-js/tree/transformExamples/integratingIntoScenesApp)

**Pull request closed without merging**